### PR TITLE
agent,api,gateway,pkg,ssh: add public url to access HTTP application on device

### PR DESCRIPTION
**Pull request by henrybarreto**

### CODE REVIEW

**Comment by henrybarreto** [comment]
Related to: https://github.com/shellhub-io/shellhub/issues/90

File: agent/main.go [hunk 1]
```diff
@@ -170,6 +170,67 @@ func NewAgentServer() *Agent {
 
 		conn.Close()
 	}
+	tun.HTTPHandler = func(w http.ResponseWriter, r *http.Request) {
+		exit := func(msg string, code int) {
```
Comment (otavio) [inline agent/main.go:174]: I'd name the `exit` function as `replyError`.

File: gateway/shellhub.conf [hunk 2]
```diff
@@ -391,3 +391,16 @@ server {
...
+   server_name ~^(?<namespace>.+)\.(?<device>.+)\..+$;
+
+   location / { #~ ^/(.*)$ {
+       rewrite ^/(.*)$ /ssh/http break;
+       proxy_set_header X-Namespace $namespace;
+       proxy_set_header X-Device $device;
+       proxy_set_header X-Path /$1$is_args$args;
+       proxy_pass http://ssh:8080;
+   }
+}
```
Comment (otavio) [inline gateway/shellhub.conf:406]: Newline at end of file.

File: pkg/api/client/client.go [hunk 3]
```diff
@@ -67,6 +67,12 @@ type client struct {
 	logger *logrus.Logger
 }
 
+func buildURL(c *client, uri string) string {
+	u, _ := url.Parse(fmt.Sprintf("%s://%s:%d%s", c.scheme, c.host, c.port, uri))
+
+	return u.String()
+}
+
```
Comment (otavio) [inline pkg/api/client/client.go:75]: Is this change intended? It seems as same code just moved to another place. I'd prefer it as another commit or PR.

File: pkg/api/client/client_public.go [hunk 4]
```diff
@@ -89,6 +89,10 @@ func (c *client) Endpoints() (*models.Endpoints, error) {
 	return endpoints, nil
 }
 
+func tunnelDial(ctx context.Context, protocol, address string, port int, path string) (*websocket.Conn, *http.Response, error) {
+	return websocket.DefaultDialer.DialContext(ctx, strings.Join([]string{fmt.Sprintf("%s://%s:%d", protocol, address, port), path}, ""), nil)
+}
+
```
Comment (otavio) [inline pkg/api/client/client_public.go:95]: same here.

File: ssh/main.go [hunk 5]
```diff
@@ -71,6 +73,65 @@ func main() {
 		}
 	})
 
+	router.HandleFunc("/ssh/http", func(w http.ResponseWriter, r *http.Request) {
+		dev, err := tunnel.API.GetDevice("d6c6a5e97217bbe4467eae46ab004695a766c5c43f70b95efd4b6a4d32b33c6e") // TODO: get device UID from request.
```
Comment (otavio) [inline ssh/main.go:77]: Fix TODO.

File: ssh/main.go [hunk 6]
```diff
@@ -71,6 +73,65 @@ func main() {
...
+		if err != nil {
+			http.Error(w, fmt.Sprintf("failed to get device: %s", err), http.StatusInternalServerError)
```
Comment (otavio) [inline ssh/main.go:79]: prepare a `replyError` function.

**Review by otavio — CHANGES_REQUESTED** [review]

**Review by otavio — DISMISSED** [review]

**Pull request merged**